src/lib.rs

```diff
@@ -108,47 +122,64 @@ pub enum FoundCrate {
+    let manifest_path = Path::new(&manifest_dir).join("Cargo.toml");
+    let manifest_ts = cargo_toml_timestamp(&manifest_path)?;
 
-    struct Cache {
-        manifest_dir: String,
-        manifest_path: PathBuf,
-        crate_names: BTreeMap<String, FoundCrate>,
-    }
-
-    static CACHE: OnceCell<Cache> = OnceCell::new();
-    let cache = CACHE.get_or_try_init(|| {
-        let manifest_dir = manifest_dir.clone();
-        let manifest_path = Path::new(&manifest_dir).join("Cargo.toml");
-
-        if !manifest_path.exists() {
-            return Err(Error::NotFound(manifest_dir.into()));
-        }
+    // This `Lazy<Mutex<_>>` can just be a `Mutex<_>` starting in Rust 1.63:
+    // https://doc.rust-lang.org/beta/std/sync/struct.Mutex.html#method.new
+    static CACHE: Lazy<Mutex<Cache>> = Lazy::new(Mutex::default);
+    let mut cache = CACHE.lock().unwrap();
 
-        let manifest = open_cargo_toml(&manifest_path)?;
-        let crate_names = extract_crate_names(&manifest)?;
+    let crate_names = match cache.entry(manifest_dir) {
+        btree_map::Entry::Occupied(entry) => {
+            let cache_entry = entry.into_mut();
 
-        Ok(Cache {
-            manifest_dir,
-            manifest_path,
-            crate_names,
-        })
-    })?;
+            // Timestamp changed, rebuild this cache entry.
+            if manifest_ts != cache_entry.manifest_ts {
+                *cache_entry = read_cargo_toml(&manifest_path, manifest_ts)?;
+            }
 
-    assert_eq!(
-        manifest_dir, cache.manifest_dir,
-        "CARGO_MANIFEST_DIR must not change within one compiler process"
-    );
+            &cache_entry.crate_names
+        }
+        btree_map::Entry::Vacant(entry) => {
+            let cache_entry = entry.insert(read_cargo_toml(&manifest_path, manifest_ts)?);
+            &cache_entry.crate_names
+        }
+    };
 
-    Ok(cache
-        .crate_names
+    Ok(crate_names
         .get(orig_name)
         .ok_or_else(|| Error::CrateNotFound {
             crate_name: orig_name.to_owned(),
-            path: cache.manifest_path.clone(),
+            path: manifest_path,
         })?
         .clone())
 }
 
+fn cargo_toml_timestamp(manifest_path: &Path) -> Result<SystemTime, Error> {
+    let timestamp = || fs::metadata(manifest_path)?.modified();
+    timestamp().map_err(|source| {
+        if source.kind() == io::ErrorKind::NotFound {
+            Error::NotFound(manifest_path.to_owned())
+        } else {
+            Error::CouldNotRead {
+                path: manifest_path.to_owned(),
+                source,
+            }
+        }
+    })
+}
+
+fn read_cargo_toml(manifest_path: &Path, manifest_ts: SystemTime) -> Result<CacheEntry, Error> {
+    let manifest = open_cargo_toml(manifest_path)?;
+    let crate_names = extract_crate_names(&manifest)?;
+
+    Ok(CacheEntry {
+        manifest_ts,
+        crate_names,
```

Comment:
Style nit: could directly assign the `crate_names` field instead of binding it separately.

Reply:
Same thing, I thinkk the separate binding makes it easier to read. Would also be shorter if rustfmt would put this struct literal on one line, which it unfortunately doesn't do in the default configuration.